### PR TITLE
Pull #13362: Restore xdoc-javadoc validation

### PR DESCRIPTION
**Pull request by stoyanK7**
Restores xdoc-javadoc validation that was removed in https://github.com/checkstyle/checkstyle/pull/13117/files#diff-350da613cecad468adf5dcb54826f8aa2cca63e699aee79a7d8bad4f8257f48d prior to splitting xdocs.

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java [hunk 1]
```diff
@@ -725,6 +725,9 @@ public class CustomImportOrderCheck extends AbstractCheck {
     /** Contains objects with import attributes. */
     private final List<ImportDetails> importToGroupList = new ArrayList<>();
 
+    /** Specify format of order declaration customizing by user. */
+    private String customImportOrderRules = "";
+
```
[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: So fun... Disabled for a bit and we leaked a problem :).

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: So we come to new nuance: there is setter, but no field.

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: @stoyanK7 , please recheck ability to get content from set method.

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: https://github.com/checkstyle/checkstyle/blob/1ef4cd7e10a71dcbc584ad306049ec355d3888de/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java#L804

Not what we need but we need decision to reuse it for xdoc content

[the remark this answers is not in the document]
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: We should not be allowing pitest to dictate class design/common patterns in Checkstyle. My expectation has always been that a user facing property has an associated field in check class (or superclass).

[the remark this answers is not in the document]
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: Wait, isn't this list made exactly for this case
https://github.com/checkstyle/checkstyle/blob/1ef4cd7e10a71dcbc584ad306049ec355d3888de/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L138-L152

This is where the test failed
https://github.com/checkstyle/checkstyle/blob/1ef4cd7e10a71dcbc584ad306049ec355d3888de/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1307-L1317

[the remark this answers is not in the document]
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: > Wait, isn't this list made exactly for this case

It looks that way, I am not sure if extending this list is going to make document generation any easier, though. For this test in this PR, sure, but my comment at https://github.com/checkstyle/checkstyle/pull/13362#discussion_r1253280784 is more philosophical.

[the remark this answers is not in the document]
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: > deprecated fields 

Can be removed, they are gone.

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: Let's use doc text from method javadoc, we can remove "Setter that" prefix to get text that we need.

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:730]: To avoid big blockages....

Let's do hack and skip this validation for this module and this property. Create separate issue to comeback to it and fix it. Let's merge this PR with hack.

We will be forced to fix it when we start to migrate/generate properties of modules.

[the remark this answers is not in the document]
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java]: Fixing CI right now.

**Review by romani — APPROVED** [review]
Ok to merge.

**Comment by romani** [comment]
`[ERROR]   XdocsJavaDocsTest.testAllCheckSectionJavaDocs:171->examineCheckSection:198 »  Error was thrown while processing src\main\java\com\puppycrawl\tools\checkstyle\checks\javadoc\SingleLineJavadocCheck.java`

@stoyanK7 , please fix windows tests.

**Comment by romani** [comment]
Still windows problem

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 2]
```diff
@@ -1385,10 +1384,15 @@ private static void validateViolationSection(String fileName, String sectionName
                 .isEqualTo("");
         }
         else {
+            final String subsectionTextContent = subSection.getTextContent()
+                    .replaceAll("\n\\s+", "\n")
+                    .replaceAll("\n", " ")
```
Comment (nrmancuso) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:1389]: Can we condense this into one `replaceAll` call, or am I missing something? `\n\\s+` -> `\n` -> ` `?

**Review by nrmancuso — CHANGES_REQUESTED** [review]
minor item form me:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java [hunk 3]
```diff
@@ -44,9 +44,10 @@
  * </p>
  * <p>
  * For additional restriction of type usage see also:
- * <a href="https://checkstyle.org/checks/coding/illegalinstantiation.html">
+ * <a href="https://checkstyle.org/checks/coding/illegalinstantiation.html#IllegalInstantiation">
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:47]: Why do we need anchors here still?

**Review by nrmancuso — COMMENTED** [review]
Just one question:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java (hunk 3 above)
Comment (stoyanK7) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:47]: There were two solutions here, either remove them from xdoc or keep them in javadoc. I remember we talked about keeping them to make it easy for mobile.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:47]: Ok, I am leaving this unresolved for next reviewer to see, but I am good.

**Review by nrmancuso — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java (hunk 3 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:47]: We will meditate on this at the end https://github.com/checkstyle/checkstyle/issues/13386

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java [hunk 4]
```diff
@@ -0,0 +1,782 @@
...
+import com.puppycrawl.tools.checkstyle.checks.whitespace.WrapOption;
+import com.puppycrawl.tools.checkstyle.internal.utils.TestUtil;
+import com.puppycrawl.tools.checkstyle.internal.utils.XdocUtil;
+import com.puppycrawl.tools.checkstyle.internal.utils.XmlUtil;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
+import com.puppycrawl.tools.checkstyle.utils.JavadocUtil;
+import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java:77]: This is not a new class. Not sure why it wasn't disabled instead of deleting.

Can you provide a diff to compare what it was before and what it is now?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 5]
```diff
@@ -553,6 +550,8 @@ public void testAllCheckSections() throws Exception {
                                     .isTrue();
                 }
 
+                validateCheckSection(moduleFactory, fileName, sectionName, section);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:553]: This and moduleFactory is just a revert of 81448ea434cda883b748d3d2c77b2f58a8b22bbe

File: src/xdocs/checks/design/finalclass.xml [hunk 6]
```diff
@@ -15,7 +15,7 @@
         </p>
         <ol>
           <li>
-              Private classes with implicit constructor.
+              Private classes with no declared constructors.
```
Comment (rnveach) [inline src/xdocs/checks/design/finalclass.xml:18]: Why was this changed? We had no issues merging this just 2 weeks ago.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 7]
```diff
@@ -1409,7 +1412,7 @@ private static void validateViolationSection(String fileName, String sectionName
                 assertWithMessage(fileName + " section '" + sectionName
                         + "' should have matching url for '" + linkText + "'")
                     .that(url)
-                    .isEqualTo(expectedUrl);
+                    .contains(expectedUrl);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:1415]: Why is this not equal to anymore? The point of `equalTo` was to ensure we have the correct URL entered, with no mistakes.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java [hunk 8]
```diff
@@ -51,6 +51,24 @@ public final class XdocUtil {
     private XdocUtil() {
     }
 
+    /**
+     * Gets all xdocs file paths. This includes Velocity .xml.vm files.
+     *
+     * @return a set of all xdocs file paths.
+     * @throws IOException if an I/O error occurs.
+     */
+    public static Set<Path> getAllXdocFilePaths() throws IOException {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java:60]: Why do we need this and not modify `getXdocsFilePaths`?

This will be confusing to people on when do they need to use "all" and "non-all" and how "non-all" is not everything and what is it dropping.

VM files are still XDoc files and should be included in the other to me.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java (hunk 8 above)
Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java:60]: I agree. This was my initial thought, but 3 tests break if I extend `getXdocsFilePaths`. I opened https://github.com/checkstyle/checkstyle/issues/13389.
Please check it out.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 7 above)
Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:1415]: Here is a failure in `testAllCheckSections` with `isEqualTo(expectedUrl)`
```
methodparampad.xml section 'MethodParamPad' should have matching url for 'see the documentation'
expected: config.html#Custom_messages
but was : ../../config.html#Custom_messages
Expected :config.html#Custom_messages
Actual   :../../config.html#Custom_messages
```

On second thought, I should have done this and not touch the `isEqualsTo()`
```diff
-                    expectedUrl = "config.html#Custom_messages";
+                    expectedUrl = "../../config.html#Custom_messages";
```

Done.

File: src/xdocs/checks/design/finalclass.xml (hunk 6 above)
Comment (stoyanK7) [reply src/xdocs/checks/design/finalclass.xml:18]: Uh, I forgot to ask about this one. The validation wasn't there 2 weeks ago so this one slipped.
@Kevin222004 Could you please tell which one is the true one(the one that should stay)?
> Private classes with implicit constructor.

or

> Private classes with no declared constructors.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java (hunk 4 above)
Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java:77]: Here you go. 2b849e426b06eb9b1777476de4c1657c71a373ba is the commit before the test was removed so this was the latest state of it.
```bash
stoyan @ legion: ~/open-source/checkstyle                                                                                                                                      pull/13361
$ git --no-pager diff 2b849e426b06eb9b1777476de4c1657c71a373ba HEAD src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java

```
```diff
diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java
index cc3964b76..c40b48aba 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java
@@ -31,6 +31,7 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -120,6 +121,8 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
 
     private static String checkName;
 
+    private static Set<Path> xdocsFilePaths;
+
     @Override
     protected String getPackageLocation() {
         return "com.puppycrawl.tools.checkstyle.internal";
@@ -130,6 +133,7 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
         final DefaultConfiguration checkConfig = new DefaultConfiguration(
                 JavaDocCapture.class.getName());
         checker = createChecker(checkConfig);
+        xdocsFilePaths = XdocUtil.getAllXdocFilePaths();
     }
 
     /**
@@ -147,7 +151,8 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
             final File file = path.toFile();
             final String fileName = file.getName();
 
-            if ("config_system_properties.xml".equals(fileName)) {
+            if ("config_system_properties.xml".equals(fileName)
+                    || "index.xml".equals(fileName)) {
                 continue;
             }
 
@@ -518,22 +523,13 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
             final String attrValue;
 
             if ("a".equals(nodeName) && "href".equals(attrName)) {
-                String value = attribute.getNodeValue();
+                final String value = attribute.getNodeValue();
 
                 assertWithMessage("links starting with '#' aren't supported: " + value)
                     .that(value.charAt(0))
                     .isNotEqualTo('#');
 
-                if (value.contains("://")) {
-                    attrValue = value;
-                }
-                else {
-                    if (value.charAt(0) == '/') {
-                        value = value.substring(1);
-                    }
-
-                    attrValue = "https://checkstyle.org/" + value;
-                }
+                attrValue = getLinkValue(value);
             }
             else {
                 attrValue = attribute.getNodeValue();
@@ -548,6 +544,57 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
         return result.toString();
     }
 
+    private static String getLinkValue(String initialValue) {
+        String value = initialValue;
+        final String attrValue;
+        if (value.contains("://")) {
+            attrValue = value;
+        }
+        else {
+            if (value.charAt(0) == '/') {
+                value = value.substring(1);
+            }
+
+            if (value.contains("../apidocs/")
+                    || value.contains("../files/")
+                    || value.contains("../styleguides")
+                    || value.contains("index.html")) {
+                value = value.replaceAll("\\.\\./", "");
+            }
+            else if (!value.contains("dtds/")) {
+                final int lastIndexOfSlash = value.lastIndexOf('/');
+                final String fileNameWithoutExtension;
+                if (lastIndexOfSlash == -1) {
+                    fileNameWithoutExtension = value.substring(0, value.indexOf(".html"));
+                }
+                else {
+                    fileNameWithoutExtension = value
+                            .substring(lastIndexOfSlash + 1, value.indexOf(".html"));
+                }
+                final Optional<Path> xdocsFilePath = xdocsFilePaths
+                        .stream()
+                        .filter(path -> {
+                            return path
+                                    .toString()
+                                    .matches(".*[\\\\/]" + fileNameWithoutExtension + "\\.xml.*");
+                        })
+                        .findFirst();
+                if (xdocsFilePath.isPresent()) {
+                    final String filePathFromRoot = xdocsFilePath.get().toString()
+                            .replaceAll("src[\\\\/]xdocs[\\\\/]", "");
+                    value = filePathFromRoot
+                            .substring(0, filePathFromRoot.lastIndexOf(".xml"))
+                            + ".html"
+                            + value.substring(value.indexOf(".html") + 5);
+                    value = value.replaceAll("\\\\", "/");
+                }
+            }
+
+            attrValue = "https://checkstyle.org/" + value;
+        }
+        return attrValue;
+    }
+
     public static class JavaDocCapture extends AbstractCheck {
         private static final Pattern SETTER_PATTERN = Pattern.compile("^set[A-Z].*");
 

```

File: src/xdocs/checks/design/finalclass.xml (hunk 6 above)
Comment (Kevin222004) [reply src/xdocs/checks/design/finalclass.xml:18]: @stoyanK7 this is the true one
`Private classes with no declared constructors`  this statment should stay.

usually I always run mvn clean verify but 

>  The validation wasn't there 2 weeks ago so this one slipped.

due to this may be i also missed

Comment (rnveach) [reply src/xdocs/checks/design/finalclass.xml:18]: Thanks. We shouldn't disable this validation unless there is some replacement.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 7 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: Its fine until it becomes an issue if you don't account for the folder depth your in.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java (hunk 8 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java:60]: So since you moved it to a new issue, is the new issue going to be handled first?

Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java:60]: No, I will handle it after this gets merged.

Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java:60]: PR is rebased

Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XdocUtil.java]: Removed new method and used `getXdocsFilePaths` instead

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 9]
```diff
@@ -1387,18 +1386,22 @@ private static void validateViolationSection(String fileName, String sectionName
                 .isEqualTo("");
         }
         else {
+            final String subsectionTextContent = subSection.getTextContent()
+                    .replaceAll("\\n\\\\s+", "\n")
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:1390]: Original regexp was `\n\\s+`. Now it is `\\n\\\\s+`, which has an extra set of `\`s.
`\\\\s+` will not match white spaces.

I am confused by this change.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java (hunk 4 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java:77]: Why is `getLinkValue` so complicated over it's original. I am definately not understand the `dtd` pathing. We have links to something under DTDs?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 9 above)
Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:1390]: This wasn't intended. I probably did it fixing https://github.com/checkstyle/checkstyle/pull/13362#discussion_r1258232047.

Reverting.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: Orig: `"\n\\s+", "\n"`
New: `"\\n\\s+", "\n"`

Shouldn't the `\\n` be `\n`?

Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: Hoenstly, I don't know. I am moving it back to `\n` just so it's consistent with others. I see both `\n` and `\\n` being used around the repository.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java (hunk 4 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java:77]: Something just seems wrong with this method unless there is more I am missing. For `../` you should be able to identify the folder you are in and determine the full path from it without having to look up through `xdocsFilePaths` or some of the other weirdness going on here.

For example, if my path is `src\xdocs\checks\design\finalclass.xml`,
`` = `checkstyle.org/checks/design/`
`../` = `checkstyle.org/checks/`
`../../` = `checkstyle.org/`

I don't need to know the full list of files involved, I just need to know where the base is.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 9 above)
Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: Yes, it should be `\n`

https://stackoverflow.com/a/59192811/9553927

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java (hunk 4 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java:77]: This looks a lot cleaner even if I don't fully know what `resolve` and `normalize` are doing.

**Comment by rnveach** [comment]
@stoyanK7 CI is failing.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java (hunk 4 above)
Comment (stoyanK7) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java:77]:
```java
value; // ../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF
value = currentXdocPath // src/xdocs/checks/whitespace/methodparampad.xml
        .getParent() // src/xdocs/checks/whitespace
        .resolve(Paths.get(value)) // src/xdocs/checks/whitespace/../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF
        .normalize() // src/xdocs/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF
        .toString()
        .replaceAll("src[\\\\/]xdocs[\\\\/]", "") // apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF
        .replaceAll("\\\\", "/");
```

**Comment by stoyanK7** [comment]
@rnveach Done.

**Review by rnveach — APPROVED** [review]
